src/coreclr-debug/activate.ts

```diff
@@ -106,50 +106,72 @@ function showDotnetToolsWarning(message: string): void {
     }
 }
 
-interface AdapterExecutableCommand {
-    command: string;
-}
+export class DebugAdapterDecriptor implements vscode.DebugAdapterDescriptorFactory {
+    public static CORECLR_DEBUG_TYPE : string = "coreclr";
+    public static CLR_DEBUG_TYPE : string = "clr";
 
-// The default extension manifest calls this command as the adapterExecutableCommand
-// If the debugger components have not finished downloading, the proxy displays an error message to the user
-// Else it will launch the debug adapter
-export async function getAdapterExecutionCommand(platformInfo: PlatformInformation, eventStream: EventStream, packageJSON: any, extensionPath: string): Promise<AdapterExecutableCommand> {
-    let util = new CoreClrDebugUtil(common.getExtensionPath());
-
-    // Check for .debugger folder. Handle if it does not exist.
-    if (!CoreClrDebugUtil.existsSync(util.debugAdapterDir())) {
-        // our install.complete file does not exist yet, meaning we have not completed the installation. Try to figure out what if anything the package manager is doing
-        // the order in which files are dealt with is this:
-        // 1. install.Begin is created
-        // 2. install.Lock is created
-        // 3. install.Begin is deleted
-        // 4. install.complete is created
-
-        // install.Lock does not exist, need to wait for packages to finish downloading.
-        let installLock = false;
-        let debuggerPackage = getRuntimeDependencyPackageWithId("Debugger", packageJSON, platformInfo, extensionPath);
-        if (debuggerPackage && debuggerPackage.installPath)
-        {
-            installLock = await common.installFileExists(debuggerPackage.installPath, common.InstallFileType.Lock);
-        }
+    private platformInfo: PlatformInformation;
+    private eventStream: EventStream;
+    private packageJSON: any;
+    private extensionPath: string;
+
+    constructor(platformInfo: PlatformInformation, eventStream: EventStream, packageJSON: any, extensionPath: string) {
+        this.platformInfo = platformInfo;
+        this.eventStream = eventStream;
+        this.packageJSON = packageJSON;
+        this.extensionPath = extensionPath;
+    }
+
+    createDebugAdapterDescriptor(session: vscode.DebugSession, executable: vscode.DebugAdapterExecutable | undefined): vscode.ProviderResult<vscode.DebugAdapterDescriptor> {
```

Comment:
This function feels complicated enough to me that we should really write it using async await instead.